src/category_theory/connected.lean

```diff
@@ -0,0 +1,374 @@
+  apply prop_up_chain' p l nemp zags hd tl _ d,
+  rintros _ _ (⟨⟨_⟩⟩ | ⟨⟨_⟩⟩),
+  apply k a, symmetry, apply k a
+end
+
+end J
+
+section examples
+instance cospan_inhabited : inhabited walking_cospan := ⟨walking_cospan.one⟩
+
+instance cospan_connected : connected (walking_cospan) :=
+begin
+  apply connected.of_induct,
+  introv _ t,
+  cases j,
+  { rwa t walking_cospan.hom.inl },
+  { rwa t walking_cospan.hom.inr },
+  { assumption }
+end
+
+instance span_inhabited : inhabited walking_span := ⟨walking_span.zero⟩
+
+instance span_connected : connected (walking_span) :=
+begin
+  apply connected.of_induct,
+  introv _ t,
+  cases j,
+  { assumption },
+  { rwa ← t walking_span.hom.fst },
+  { rwa ← t walking_span.hom.snd },
+end
+
+instance parallel_pair_inhabited : inhabited walking_parallel_pair := ⟨walking_parallel_pair.one⟩
+
+instance parallel_pair_connected : connected (walking_parallel_pair) :=
+begin
+  apply connected.of_induct,
+  introv _ t, cases j,
+  { rwa t walking_parallel_pair_hom.left },
+  { assumption }
+end
+
+end examples
+
+section C
+variables {J : Type v₂} [𝒥 : category.{v₁} J]
+include 𝒥
+
+variables {C : Type u₂} [𝒞 : category.{v₂} C]
+include 𝒞
+
+/--
+For objects `X Y : C`, any natural transformation `α : const X ⟶ const Y` from a connected
+category must be constant.
+This is the key property of connected categories which we use to establish properties about limits.
+-/
+lemma nat_trans_from_connected [conn : connected J] {X Y : C}
+  (α : (functor.const J).obj X ⟶ (functor.const J).obj Y) :
+  ∀ (j : J), α.app j = (α.app (default J) : X ⟶ Y) :=
+@constant_function_of_preserves_morphisms _ _ _
+  (X ⟶ Y)
+  (λ j, α.app j)
+  (λ _ _ f, (by { have := α.naturality f, erw [id_comp, comp_id] at this, exact this.symm }))
+
+end C
+
+local attribute [tidy] tactic.case_bash
+
+variables {C : Type u₂} [𝒞 : category.{v₂} C]
+include 𝒞
+
+section products
+
+variables [has_binary_products.{v₂} C]
+
+variables {J : Type v₂} [small_category J]
+
+/-- (Impl). The obvious natural transformation from (X × K -) to K. -/
+@[simps]
+def γ₂ {K : J ⥤ C} (X : C) : K ⋙ prod_functor.obj X ⟶ K :=
+{ app := λ Y, limits.prod.snd }
+
+/-- (Impl). The obvious natural transformation from (X × K -) to X -/
+@[simps]
+def γ₁ {K : J ⥤ C} (X : C) : K ⋙ prod_functor.obj X ⟶ (functor.const J).obj X :=
+{ app := λ Y, limits.prod.fst }
+
+/-- (Impl). Given a cone for (X × K -), produce a cone for K using the natural transformation `γ₂` -/
+@[simps]
+def forget_cone {X : C} {K : J ⥤ C} (s : cone (K ⋙ prod_functor.obj X)) : cone K :=
+{ X := s.X,
+  π := s.π ≫ γ₂ X }
+
+/--
+The functor `(X × -)` preserves any connected limit.
+Note that this functor does not preserve the two most obvious disconnected limits - that is,
+`(X × -)` does not preserve products or terminal object, eg `(X ⨯ A) ⨯ (X ⨯ B)` is not isomorphic to
+`X ⨯ (A ⨯ B)` and `X ⨯ 1` is not isomorphic to `1`.
+-/
+def prod_preserves_connected_limits [connected J] (X : C) :
```

Comment:
Can you combine successive rewrites into `rw [...]`? I appreciate it's a hassle here because we need `erw` as well as `rw`. I think it's okay to just combine them into a `erw [...]`.

I'm still thinking about ways to diagnose the necessity for `erw`, which is very unfortunate. :-(